core/command_parser.h

```diff
@@ -0,0 +1,246 @@
+	int _validate_arguments(int p_current_idx); // Returns number of arguments taken, -1 on validation error.
+
+	// Helpers for the function above that parse a specific case
+	int _validate_positional(const String &p_arg, int p_current_idx);
+	int _validate_adjacent(const String &p_arg, const String &p_prefix, int p_separator);
+	int _validate_short(const String &p_arg, const String &p_prefix, int p_current_idx);
+	int _validate_long(const String &p_arg, const String &p_prefix, int p_current_idx);
+
+	// Validation helpers
+	const CommandOption *_validate_command(const String &p_name, const String &p_prefix);
+	int _validate_command_args(const CommandOption *p_command, const String &p_display_name, int p_current_idx, bool p_skip_first = false);
+	bool _validate_command_arg(const CommandOption *p_command, const String &p_display_name, const String &p_arg);
+
+	// Save information about parsed command
+	void _save_parsed_command(const CommandOption *p_command, const String &p_prefix, int p_idx, int p_args_count, const String &p_additional_value = String());
+	void _save_parsed_command(const CommandOption *p_command, const String &p_prefix, const String &p_value = String());
+	void _save_parsed_command(const CommandOption *p_command, int p_idx, int p_args_count);
+
+	// Help text printers
+	String _get_usage(const Vector<Pair<const CommandOption *, String>> &p_printable_commands) const;
+	String _get_commands_description(const OrderedHashMap<String, PackedStringArray> &p_categories_data) const;
+
+	// Other utilies
+	String _to_string(const PackedStringArray &p_names) const; // Returns all command names separated by commas with all prefix variants.
+	String _get_prefixed_longest_name(const PackedStringArray &p_names) const; // Returns longest name with first available prefix (long or short).
+	ParsedPrefix _parse_prefix(const String &p_arg) const;
+	String _find_most_similar(const String &p_name) const;
+	static bool _contains_optional_commands(const Vector<Pair<const CommandOption *, String>> &p_printable_commands);
+
+protected:
+	static void _bind_methods();
+
+public:
+	Error parse_args(const PackedStringArray &p_args);
+	Error notify();
+
+	void add_command(const Ref<CommandOption> &p_command);
+	int get_commands_count() const;
+	Ref<CommandOption> get_command(int p_idx) const;
+	void set_command(int p_idx, const Ref<CommandOption> &p_command);
+	void remove_command(int p_idx);
+	Ref<CommandOption> find_command(const String &p_name) const;
+
+	Ref<CommandOption> add_help_command();
+	Ref<CommandOption> add_version_command();
+
+	bool is_set(const Ref<CommandOption> &p_command) const;
+	String get_value(const Ref<CommandOption> &p_command) const;
+	PackedStringArray get_values(const Ref<CommandOption> &p_command) const;
+
+	String get_prefix(const Ref<CommandOption> &p_command) const;
+	PackedStringArray get_prefixes(const Ref<CommandOption> &p_command) const;
+
+	int get_count(const Ref<CommandOption> &p_command) const;
```

Comment:
Totally agree.

Reply:
Hm... I think that the `_option` part in name is redundant because it obvious from the argument. And `occurences` does not mean that the function returns a count. I'd stick with the previous name.

Reply:
> I'd use "option" because that's a noun which should be taken as a base for the entire API.

Renamed all `command` usages to `option`!